File: 89_Tic-Tac-Toe/java/src/Board.java
/**
 * @author Ollie Hensman-Crook
 */
public class Board {
    private char arr[];

    public Board() {
        this.arr = new char[9];
        for (int x = 1; x <= 9; x++) {
            this.arr[x - 1] = ' ';
        }
    }


    /**
     * Place 'X' or 'O' on the board position passed
     * @param position
     * @param player
     */
    public void setArr(int position, char player) {
        if (player == 'X') {
            this.arr[position-1] = 'X';
        } else {
            this.arr[position -1] = 'O';
        }
    }

    public void printBoard() {
        System.out.format("%-3c ! %-3c ! %-3c\n----+----+----\n%-3c ! %-3c ! %-3c\n----+----+----\n%-3c ! %-3c ! %-3c\n",
        this.arr[0], this.arr[1], this.arr[2], this.arr[3], this.arr[4], this.arr[5], this.arr[6], this.arr[7], this.arr[8]
        );
    }


    /**
     * @param x
     * @return the value of the char at a given position
     */
    public char getBoardValue(int x) {
        return arr[x-1];
    }


    /**
     * Go through the board and check for win (horizontal, diagonal, vertical)
     * @param player
     * @return whether a win has occured
     */
    public boolean checkWin(char player) {
        if(this.arr[0] == player && this.arr[1] == player && this.arr[2] == player)
            return true;


        if(this.arr[3] == player && this.arr[4] == player && this.arr[5] == player)
            return true;


        if(this.arr[6] == player && this.arr[7] == player && this.arr[8] == player)
            return true;

        if(this.arr[0] == player && this.arr[4] == player && this.arr[8] == player)
            return true;

        if(this.arr[2] == player && this.arr[4] == player && this.arr[6] == player)
            return true;

        if(this.arr[0] == player && this.arr[3] == player && this.arr[6] == player)
            return true;

        if(this.arr[1] == player && this.arr[4] == player && this.arr[7] == player)
            return true;

        if(this.arr[2] == player && this.arr[5] == player && this.arr[8] == player)
            return true;

        return false;
    }
    public boolean checkDraw() {
        if(this.checkWin('X') == false && this.checkWin('O') == false) {
            if(this.getBoardValue(1) != ' ' && this.getBoardValue(2) != ' ' && this.getBoardValue(3) != ' ' && this.getBoardValue(4) != ' ' && this.getBoardValue(5) != ' ' && this.getBoardValue(6) != ' ' && this.getBoardValue(7) != ' ' && this.getBoardValue(8) != ' ' && this.getBoardValue(9) != ' ' ) {
                return true;
            }
        }

        return false;
    }
    /**
     * Reset the board
     */
    public void clear() {
        for (int x = 1; x <= 9; x++) {
            this.arr[x - 1] = ' ';
        }
    }

}

File: 89_Tic-Tac-Toe/kotlin/Board.kt
/**
 * @author John Long based on Java by Ollie Hensman-Crook
 */

enum class Player(val char: Char) {
    X('X'),
    O('O')
}

class Board {
    // Initialize an array of size nine with all values set to null
    private var boxes: Array<Player?> = arrayOfNulls(9)

    /**
     * Place 'X' or 'O' on the board position passed
     * @param position
     * @param player
     */
    fun setArr(position: Int, player: Player) {
        boxes[position - 1] = player
    }

    fun printBoard() {
        System.out.format(
            """
  %c !  %c !  %c
----+----+----
  %c !  %c !  %c
----+----+----
  %c !  %c !  %c
""",
            // converts each box to a char and then passes them in order to format
            // If the person is unassigned, use a space ' '
     *(boxes.map{it?.char ?: ' '}.toTypedArray()))
    }

    /**
     * @param x
     * @return the value of the char at a given position
     */
    fun getBoardValue(x: Int): Player? {
        return boxes[x - 1]
    }

    private val winningCombos = listOf(
        // horizontal
        listOf(0,1,2),
        listOf(3,4,5),
        listOf(6,7,8),
        // diagonal
        listOf(0,4,8),
        listOf(2,4,6),
        // vertical
        listOf(0,3,6),
        listOf(1,4,7),
        listOf(2,5,8)
    )
    /**
     * Go through the board and check for win
     * @param player
     * @return whether a win has occurred
     */
    fun isWinFor(player: Player): Boolean {
        // Check if any winningCombos have all their boxes set to player
        return winningCombos.any{ combo ->
            combo.all { boxes[it] == player }
        }
    }

    fun isDraw(): Boolean {
        return !isWinFor(Player.X) && !isWinFor(Player.O) && boxes.all { it != null }
    }

    /**
     * Reset the board
     */
    fun clear() {
        boxes = arrayOfNulls(9)
    }
}
